src/lib.rs

```diff
@@ -1,7 +1,168 @@
+//! Traits and Structs for vhost-based virtio devices.
+//!
+//! Virtio devices use virtqueues to transport data efficiently. Virtqueue is a set of three
+//! different single-producer, single-consumer ring structures designed to store generic
+//! scatter-gatter I/O.
+//!
+//! Vhost is a protocol for devices accessible via inter-process communication. It uses the same
+//! virtqueue layout as Virtio to allow Vhost devices to be mapped directly to Virtio devices.
+//! This allows a Vhost device to be accessed directly by a guest OS inside a hypervisor process
+//! with an existing Virtio (PCI) driver. Only the configuration, I/O submission notification,
+//! and I/O completion interruption are piped through the hypervisor.
+//!
+//! The initial vhost implementation is a part of the Linux kernel and uses ioctl interface to
+//! communicate with userspace applications. Later Vhost-user protocol is introduced to complement
+//! the ioctl interface used to control the vhost implementation in the Linux kernel. It implements
+//! the control plane needed to establish virtqueue sharing with a user space process on the same
+//! host. It uses communication over a Unix domain socket to share file descriptors in the
+//! ancillary data of the message. The protocol defines 2 sides of the communication, master and
+//! slave. Master is the application that shares its virtqueues. Slave is the consumer of the
+//! virtqueues. Master and slave can be either a client (i.e. connecting) or server (listening)
+//! in the socket communication.
+
+#![deny(missing_docs)]
+
+#[cfg(feature = "vhost-device")]
+extern crate sys_util;
+extern crate vm_memory;
+
+#[cfg(feature = "vhost-device")]
+use sys_util::EventFd;
+#[cfg(feature = "vhost-device")]
+use vm_memory::{GuestAddress, GuestMemory};
+
+/// Error codes associated with handling vHost commands.
+#[derive(Debug)]
+pub enum Error {
+    /// Invalid operations.
+    InvalidOperation,
+    /// Invalid guest memory.
+    InvalidGuestMemory,
+    /// Invalid guest memory region.
+    InvalidGuestMemoryRegion,
+    /// Invalid queue.
+    InvalidQueue,
+    /// Invalid descriptor table address.
+    DescriptorTableAddress,
+    /// Invalid used address.
+    UsedAddress,
+    /// Invalid available address.
+    AvailAddress,
+    /// Invalid log address.
+    LogAddress,
+    /// Error opening the vhost backend device.
+    VhostOpen(std::io::Error),
+    /// Error while running ioctl.
+    IoctlError(std::io::Error),
+}
+
+/// Result of vHost operations
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// An interface for setting up vhost-based virtio devices.
+///
+/// Vhost-based virtio devices are different from regular virtio devices because the the vhost
+/// backend takes care of handling all the data transfer. The device itself only needs to deal with
+/// setting up the the backend driver and managing the control channel.
+#[cfg(feature = "vhost-device")]
+pub trait VhostDevice<'a>: std::marker::Sized {
```

Comment:
Thinking about the future with `vhost-user`, do you expect `vhost-user` devices to implement this trait directly (given that it needs to be extended)? Or do you expect the creation of a new `trait` `VhostUserDevice` that would inherit from this trait?

Reply:
I think it would be better that vhost-user implements the VhostDevice trait, though we need to extend the VhostDevice trait. Otherwise VhostDevice trait is almost useless:)

Reply:
Yes, I would expect at least two implementations for vhost-kernel and vhost-user respectively.

Reply:
`VhostDevice` trait is needed if you want to implement both `vhost_vsock` and `vhost_net`, right?

But in case of vhost-user, I'm think we should create a trait extending `VhostDevice`, as it will be cleaner and clearer from any vhost-user device to implement a trait specific to vhost-user.

Reply:
Maybe rename this to VhostBackend?